rhino/src/main/java/org/mozilla/javascript/ClassCache.java

```diff
@@ -137,16 +137,24 @@ public synchronized void setCachingEnabled(boolean enabled) {
      */
     Map<CacheKey, JavaMembers> getClassCacheMap() {
         if (classTable == null) {
-            // Use 1 as concurrency level here and for other concurrent hash maps
-            // as we don't expect high levels of sustained concurrent writes.
-            classTable = new ConcurrentHashMap<>(16, 0.75f, 1);
+            synchronized (this) {
+                if (classTable == null) {
+                    // Use 1 as concurrency level here and for other concurrent hash maps
+                    // as we don't expect high levels of sustained concurrent writes.
+                    classTable = new ConcurrentHashMap<>(16, 0.75f, 1);
```

Comment:
Just as a general principle -- do we have a sense that it actually makes sense for us to hard-code our own values for cache size, ratio, etc? From time to time the java libraries change the defaults as they learn more. I would be happier if we'd just use the default constructor here and elsewhere unless we have actual data.

Reply:
yes, especially in this case, the implementation of ConcurrentHashMap seems to be completely rewritten from JDK7 to 8

https://github.com/openjdk/jdk7/blob/master/jdk/src/share/classes/java/util/concurrent/ConcurrentHashMap.java#L735
https://github.com/openjdk/jdk8/blob/master/jdk/src/share/classes/java/util/concurrent/ConcurrentHashMap.java#L892

There is still https://github.com/mozilla/rhino/blob/master/rhino/src/main/java/org/mozilla/javascript/commonjs/module/provider/CachingModuleScriptProviderBase.java#L34 which seems to have adopted some code from JDK6 ConcurrentHashMap and still uses the individual constructors